buildSrc/src/main/java/net/neoforged/neodev/GenerateAccessTransformers.java

```diff
@@ -0,0 +1,182 @@
+        var text = new StringBuilder();
+        for (int i = 0; i < groups.length; i++) {
+            // Check if the group found no targets. If it didn't, there's probably an error in the test and it should be reported
+            if (groups[i].isEmpty()) {
+                throw new IllegalStateException("Generated AT group '" + groupList.get(i).name + "' found no entries!");
+            }
+            text.append("# ").append(groupList.get(i).name).append('\n');
+            text.append(groups[i].stream().sorted().collect(Collectors.joining("\n")));
+            text.append('\n');
+
+            if (i < groups.length - 1) text.append('\n');
+        }
+
+        var outFile = getAccessTransformer().getAsFile().get().toPath();
+        if (!Files.exists(outFile.getParent())) {
+            Files.createDirectories(outFile.getParent());
+        }
+
+        FileUtils.writeStringSafe(outFile, text.toString(), StandardCharsets.UTF_8);
+    }
+
+    public void classGroup(String name, Modifier modifier, SerializablePredicate<ClassInfo> match) {
+        getGroups().add(new AtGroup(name, modifier, match, null, null));
+    }
+
+    public void methodGroup(String name, Modifier modifier, SerializablePredicate<ClassInfo> targetTest, SerializablePredicate<MethodInfo> methodTest) {
+        getGroups().add(new AtGroup(name, modifier, targetTest, methodTest, null));
+    }
+
+    public void fieldGroup(String name, Modifier modifier, SerializablePredicate<ClassInfo> targetTest, SerializablePredicate<FieldInfo> fieldTest) {
+        getGroups().add(new AtGroup(name, modifier, targetTest, null, fieldTest));
+    }
+
+    public <T extends Named> SerializablePredicate<T> named(String name) {
+        return target -> target.getName().equals(name);
+    }
+
+    public SerializablePredicate<ClassInfo> classesWithSuperclass(String superClass) {
+        return target -> target.hasSuperclass(superClass);
+    }
+
+    public SerializablePredicate<ClassInfo> innerClassesOf(String parent) {
+        var parentFullName = parent + "$";
+        return target -> target.name().startsWith(parentFullName);
+    }
+
+    public SerializablePredicate<MethodInfo> methodsReturning(String type) {
+        var endMatch = ")L" + type + ";";
+        return methodInfo -> methodInfo.descriptor().endsWith(endMatch);
+    }
+
+    public SerializablePredicate<FieldInfo> fieldsOfType(SerializablePredicate<ClassInfo> type) {
+        return value -> type.test(value.type());
+    }
+
+    public <T> SerializablePredicate<T> matchAny() {
+        return value -> true;
+    }
+
+    public record AtGroup(String name, Modifier modifier, SerializablePredicate<ClassInfo> classMatch,
+                          @Nullable SerializablePredicate<MethodInfo> methodMatch, @Nullable SerializablePredicate<FieldInfo> fieldMatch) implements Serializable {
+    }
+
+    public record Modifier(String name, boolean isFinal, int... validOpcodes) implements Serializable, IntPredicate {
```

Comment:
Do you use this as an IntPredicate? If not, remove the impl.

Reply:
Yes, the test method is used to check if the modifier matches the one already present.

Reply:
Yes but you don't need to implement `IntPredicate`?